Source/Core/DolphinWX/LuaThread.cpp

```diff
@@ -0,0 +1,66 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "LuaScripting.h"
+
+#include "Core\Core.h"
+
+namespace Lua
+{
+GCPadStatus* pad_status;
+
+LuaThread::LuaThread(LuaScriptFrame* p, wxString file) : wxThread()
+{
+  parent = p;
+  file_path = file;
+
+  // Initialize virtual controller
+  pad_status = (GCPadStatus*)malloc(sizeof(GCPadStatus));
+  ClearPad(pad_status);
+}
+
+LuaThread::~LuaThread()
+{
+  // Delete virtual controller
+  std::unique_lock<std::mutex> lock(lua_mutex);
+  free(pad_status);
+  pad_status = nullptr;
+
+  parent->NullifyLuaThread();
+}
+
+wxThread::ExitCode LuaThread::Entry()
+{
+  // Pause emu
+  Core::SetState(Core::State::Paused);
+
+  lua_State* state = luaL_newstate();

```

Comment:
```cpp
std::unique_ptr<lua_State, decltype(&lua_close)> state{luaL_newstate(), lua_close};
```

To automatically handle leak cases. For example in the early return cases below you'd leak state memory, since `lua_close` would never be called.

Reply:
My understanding of a unique_ptr is that the object referenced by the pointer can only be referenced by exactly one pointer, and as a consequence, you can't pass a unique_ptr into a function. lua_State is passed into many functions, so I'm not sure how that's supposed to work.

I see what you mean about memory leaking, but my instinct is to just call lua_close at the end of each branch,

What's so wrong with just having a normal pointer?

Reply:
You can call `std::unique_ptr::get()` to pass the pointer to functions that don't *own* the lua_State -- which is the case for all of the functions you're passing the state to.

Of course you can just call lua_close every time in a failure path, but that's unnecessary code duplication and it's error prone. With a unique_ptr, you can ensure that the state is freed without having to care about it.

Reply:
@leoetlino pretty much nailed most of it. `std::unique_ptr` doesn't represent a single reference, it represents single ownership (meanwhile `std::shared_ptr` and `std::weak_ptr` are used in tandem to represent shared ownership by reference count). It makes lifetime and ownership responsibilities explicit. In modern (C++11 and up), raw pointers have come to generally signify 'non-owning' pointers in most cases.

e.g.

```cpp
bool SomeFunction(const Type* non_owning)
{
  if (!non_owning)
    return false;

  // Do things with 'non_owning' here...

  return true;
} 


void CallingFunction()
{
  std::unique_ptr<Type> ptr = // ... some allocation function (or call to std::make_unique)

  // Some other calls that initialize and set info in Type ...

  // No matter what happens here...

  SomeFunction(ptr.get());

  // Or here, memory is guaranteed to be deallocated when the function exits,
  // instead of needing to slap deallocation calls explicitly in every code path
  // that exits the function.
}

Reply:
Ok, and since SomeFunction doesn't own the pointer it isn't automatically deallocated afer SomeFunction returns, but it IS automatically deallocated after CallingFunction returns because it owns the pointer.